apps/survey/contracts/Survey.sol

```diff
@@ -0,0 +1,250 @@
+        // (as stated before, index 0 is for NO_VOTE option)
+        survey.voters[msg.sender].castedVotes[0].stake = voterStake.sub(totalVoted);
+
+        // register number of options voted
+        survey.voters[msg.sender].optionsCastedLength = _options.length;
+
+        // add voter tokens to participation
+        survey.participation = survey.participation.add(totalVoted);
+    }
+
+    /**
+    * @notice Vote option #`_option` in survey #`_surveyId`.
+    * @dev It will use the whole balance. If voting option is 0, that's like cancelling the previously casted vote
+    * @param _surveyId Id for survey
+    * @param _option Index of supported option
+    */
+    function voteOption(uint256 _surveyId, uint256 _option) public {
+        SurveyStruct storage survey = surveys[_surveyId];
+        // this could re-enter, though we can asume the governance token is not maliciuous
+        uint256 voterStake = token.balanceOfAt(msg.sender, survey.snapshotBlock);
+        uint256[] memory options = new uint256[](1);
+        uint256[] memory stakes = new uint256[](1);
+        options[0] = _option;
+        stakes[0] = voterStake;
+
+        voteOptions(_surveyId, options, stakes);
+    }
+
+    function canVote(uint256 _surveyId, address _voter) public view returns (bool) {
+        SurveyStruct storage survey = surveys[_surveyId];
+
+        return _isSurveyOpen(survey) && token.balanceOfAt(_voter, survey.snapshotBlock) > 0;
+    }
+
+    function getSurvey(uint256 _surveyId) public view returns (bool open, address creator, uint64 startDate, uint256 snapshotBlock, uint256 minParticipationPct, uint256 votingPower, uint256 participation, uint256 options) {
+        SurveyStruct storage survey = surveys[_surveyId];
+
+        open = _isSurveyOpen(survey);
+        creator = survey.creator;
+        startDate = survey.startDate;
+        snapshotBlock = survey.snapshotBlock;
+        minParticipationPct = survey.minAcceptParticipationPct;
+        votingPower = survey.votingPower;
+        participation = survey.participation;
+        options = survey.options;
+    }
+
+    function getSurveyMetadata(uint256 _surveyId) public view returns (string) {
+        return surveys[_surveyId].metadata;
+    }
+
+    function getVoterState(uint256 _surveyId, address _voter) public view returns (uint256[] options, uint256[] stakes) {
+        if (surveys[_surveyId].voters[_voter].optionsCastedLength == 0) {
+            return (new uint256[](0), new uint256[](0));
+        }
+
+        options = new uint256[](surveys[_surveyId].voters[_voter].optionsCastedLength + 1);
+        stakes = new uint256[](surveys[_surveyId].voters[_voter].optionsCastedLength + 1);
+        for (uint256 i = 0; i <= surveys[_surveyId].voters[_voter].optionsCastedLength; i++) {
+            options[i] = surveys[_surveyId].voters[_voter].castedVotes[i].optionId;
+            stakes[i] = surveys[_surveyId].voters[_voter].castedVotes[i].stake;
```

Comment:
It might be nice to use a variable for this particular vote, so it's not getting repeated all the time, e.g.

```sol
MultiOptionVote storage vote = surveys[_surveyId].voters[_voter]
```